### DISCUSSION TITLE
Proposal: Source modifying files - for source generation

### DISCUSSION THREAD

**Discussion by ymassad** [dotnet/csharplang#3566] [Language Ideas]
Let me explain the proposal with an example. Let's say we have a file called SourceFile.cs with the following content (I also added line numbers):
```
00 public class Person
01 {
02     public string Name {get;}
03 }
```
Then, via the proposed feature, you can add a Modifier1.mod file with the following content:
```
<Modifier ForFile="SourceFile.cs">
    <AddText Line=2 Column=29 Text="\r\n    public Person(string name) => Name = name;"/>
</Modifier>
```

Ths syntax of the mod file I put here is just from the top of my head. There might be much better syntaxes. I just want to explain the general idea.

This will not cause the SourceFile.cs file to be modified. However, when compiling, the compiler will process the "Modifier" files and "modify" the source files (not on disk so not visible in Visual Studio) before actually compiling the files. So the syntax trees that will be given to the compiler will contain the "modified" versions of the files.

In the example above, the "modified" SourceFile.cs will contain:
```
public class Person
{
    public string Name {get;}
    public Person(string name) => Name = name;
}
```
Tooling can be updated to show the added texts in the source files in the editor (e.g. via adornments in VS?). Of course, these pieces will be read only in the source files(in the editor, they don't exist in the source files). The developers can choose to show or hide these read-only texts.

The main use case of this feature is source generation. It will enable much more scenarios.

Source generator developers can generate these .mod files.

**Comment by svick** [comment]
Source generators (in quite a different form than what you're proposing) are already under development. See [*Introducing C# Source Generators*](https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/).

**Comment by HaloFour** [comment]
The original design of source generators allowed for the generator to modify the semantics of existing code, and that was something that the team explicitly rejected.

To accomplish what you want you cannot affect existing source, but if `Person` is declared as `partial` you can have a source generator emit a separate part that can add members.

**Comment by ymassad** [comment]
@svick , @HaloFour , this proposed feature is complementary to source generators.

And via source generators, developers will generate .mod files, not modify the source .cs files.

**Comment by HaloFour** [comment]
You're proposing that `.mod` files can modify existing source files as it's being compiled, and no, that will not be possible.  That's not complementary, that's very much against the design that the team wants for source generators.

You can certainly write a source generator that would use an external file to influence the source that it generates, but it will not be able to affect any existing source, you can only add new source.

**Comment by ymassad** [comment]
One could argue that "partial" is actually a way to modify source files. But it is a limited way. It enables you to "modify" a class to add a new member. But it is limited in a way that it does not allow you to add a new line to an existing member.

**Comment by ymassad** [comment]
By the way, this feature can be limited in a way to enable addition only. That is, it can be implemented in a way where only "AddText" xml elements can be used.

**Comment by Joe4evr** [comment]
> But it is limited in a way that it does not allow you to add a new line to an existing member.

Yes, and Source Generators were explicitly designed that way because modifying existing source code makes for a terrible IDE experience like debugging, and problems with multiple generators trying to modify the same piece of code.

**Comment by HaloFour** [comment]
@ymassad 

I'd suggest checking out the preview release of source generators.  That's likely very close to what will ship with C# 9.0 later this year, assuming no delays.  Even if there are delays the design is very, very unlikely to change much at all.  Writing a source generator is not going to involve authoring XML files, nor is there going to be a way to change the source of a particular file before the compiler parses it (the generator has to be able to inspect the AST to know what to generate afterall).  Also, notably, source generators are a feature of the Roslyn compiler, not of the C# language.  There are no language features that are being added specifically to enable source generators.
